From patchwork Wed Mar 31 02:30:24 2021
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
X-Patchwork-Submitter: Sean Christopherson <seanjc@google.com>
X-Patchwork-Id: 12174151
Return-Path: <kvm-owner@kernel.org>
X-Spam-Checker-Version: SpamAssassin 3.4.0 (2014-02-07) on
	aws-us-west-2-korg-lkml-1.web.codeaurora.org
X-Spam-Level: 
X-Spam-Status: No, score=-26.3 required=3.0 tests=BAYES_00,DKIMWL_WL_MED,
	DKIM_SIGNED,DKIM_VALID,DKIM_VALID_AU,HEADER_FROM_DIFFERENT_DOMAINS,
	INCLUDES_CR_TRAILER,INCLUDES_PATCH,MAILING_LIST_MULTI,SPF_HELO_NONE,SPF_PASS,
	USER_AGENT_GIT,USER_IN_DEF_DKIM_WL autolearn=ham autolearn_force=no
	version=3.4.0
Received: from mail.kernel.org (mail.kernel.org [198.145.29.99])
	by smtp.lore.kernel.org (Postfix) with ESMTP id 2C37EC433DB
	for <kvm@archiver.kernel.org>; Wed, 31 Mar 2021 02:31:34 +0000 (UTC)
Received: from vger.kernel.org (vger.kernel.org [23.128.96.18])
	by mail.kernel.org (Postfix) with ESMTP id E0FEA619BB
	for <kvm@archiver.kernel.org>; Wed, 31 Mar 2021 02:31:33 +0000 (UTC)
Received: (majordomo@vger.kernel.org) by vger.kernel.org via listexpand
        id S233240AbhCaCbC (ORCPT <rfc822;kvm@archiver.kernel.org>);
        Tue, 30 Mar 2021 22:31:02 -0400
Received: from lindbergh.monkeyblade.net ([23.128.96.19]:59908 "EHLO
        lindbergh.monkeyblade.net" rhost-flags-OK-OK-OK-OK) by vger.kernel.org
        with ESMTP id S233069AbhCaCab (ORCPT <rfc822;kvm@vger.kernel.org>);
        Tue, 30 Mar 2021 22:30:31 -0400
Received: from mail-yb1-xb49.google.com (mail-yb1-xb49.google.com
 [IPv6:2607:f8b0:4864:20::b49])
        by lindbergh.monkeyblade.net (Postfix) with ESMTPS id BECD4C06175F
        for <kvm@vger.kernel.org>; Tue, 30 Mar 2021 19:30:31 -0700 (PDT)
Received: by mail-yb1-xb49.google.com with SMTP id o129so646823ybg.23
        for <kvm@vger.kernel.org>; Tue, 30 Mar 2021 19:30:31 -0700 (PDT)
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=google.com; s=20161025;
        h=reply-to:date:in-reply-to:message-id:mime-version:references
         :subject:from:to:cc;
        bh=uV4x6+EdkMXZhg7wMEUNEc/cXMhh5NFoJhoLXAZa4XA=;
        b=hyxAm79V1CoNeCcaWjPRj6j5vKrjB66hHPXwlccvH/3YB10lOq1utFmsZxYp8F3HpG
         WOVVmSaroJAJOzz4lNDIbz7ei9dCL3cKP4c17RXAj462cjZTMs8oecds+x1bGGRaLRVE
         vmKYsAvI44f3RBbPqlZk3+kcxA5yCOsa4zYDyKO9Y6bRuhE8JVlAQhJa+kJ+6Wv/2Ghz
         4xMYiQHQSD02fcexUHEgrkqc/uqoNUwmNr9s/ylUQx15EFvA2VaXTuppXEfG0Eg2XXWz
         B2M6KfrdkALt0ZTB/hRS7drJULF5z50LaOIqMSYoWDCtbYm31kQwZ7CcI9NMzcJM3792
         c7fw==
X-Google-DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=1e100.net; s=20161025;
        h=x-gm-message-state:reply-to:date:in-reply-to:message-id
         :mime-version:references:subject:from:to:cc;
        bh=uV4x6+EdkMXZhg7wMEUNEc/cXMhh5NFoJhoLXAZa4XA=;
        b=lmGLfYDLavfy7gaEUnAvl6FN7lABcK2nA2nNYGiNUR4XCd5TGgzGWclz4wicK/GRE5
         Eu+vGWFbW5TNpcnUrMLNMBoLf25pPt7FOO/smbztMc6LN3B+iuokijY2R7DX4b0lms9T
         NxABHKIjPWP0z+pzmQDAhU57OCOdYlyZDAGYB1yziHH03gBKVJPKLnFIM5+4m/RXgSD9
         v5daLAA77gXD3CVLxm84t9eqcdPbz8K9SYZJ/13KMDdDJGlTIe3jSP1QrfIuEWa1g+YS
         h8hZLeqoTI+zsif/98B0vRU8I/RFzgrUcL+XYe94HgaZ31agbc2yeBbXaNMMA9pjEGRw
         Dd2A==
X-Gm-Message-State: AOAM532bbf4YAueqHvJ+z4tDhIgJaTifb6+rtuRdYiciFs6vI/QIZAG+
        kgT3mLvyi5AnXPERdeFdG5IPVRIJ86E=
X-Google-Smtp-Source: 
 ABdhPJyoJU2Rdi/to8R+u7+o3IMo1VtUs59XwyVJUPXLriy3TmcY1tCAqdD1U35l7io8CdK4mvxxJmJF+eM=
X-Received: from seanjc798194.pdx.corp.google.com
 ([2620:15c:f:10:6c6b:5d63:9b3b:4a77])
 (user=seanjc job=sendgmr) by 2002:a25:c588:: with SMTP id
 v130mr1592057ybe.312.1617157831002;
 Tue, 30 Mar 2021 19:30:31 -0700 (PDT)
Reply-To: Sean Christopherson <seanjc@google.com>
Date: Tue, 30 Mar 2021 19:30:24 -0700
In-Reply-To: <20210331023025.2485960-1-seanjc@google.com>
Message-Id: <20210331023025.2485960-2-seanjc@google.com>
Mime-Version: 1.0
References: <20210331023025.2485960-1-seanjc@google.com>
X-Mailer: git-send-email 2.31.0.291.g576ba9dcdaf-goog
Subject: [PATCH 1/2] KVM: Account memory allocations for 'struct kvm_vcpu'
From: Sean Christopherson <seanjc@google.com>
To: Paolo Bonzini <pbonzini@redhat.com>
Cc: Sean Christopherson <seanjc@google.com>,
        Vitaly Kuznetsov <vkuznets@redhat.com>,
        Wanpeng Li <wanpengli@tencent.com>,
        Jim Mattson <jmattson@google.com>,
        Joerg Roedel <joro@8bytes.org>, kvm@vger.kernel.org,
        linux-kernel@vger.kernel.org
Precedence: bulk
List-ID: <kvm.vger.kernel.org>
X-Mailing-List: kvm@vger.kernel.org

Use GFP_KERNEL_ACCOUNT for the vCPU allocations, the vCPUs are very much
tied to a single task/VM.  For x86, the allocations were accounted up
until the allocation code was moved to common KVM.  For all other
architectures, vCPU allocations were never previously accounted, but only
because most architectures lack accounting in general (for KVM).

Fixes: e529ef66e6b5 ("KVM: Move vcpu alloc and init invocation to common code")
Signed-off-by: Sean Christopherson <seanjc@google.com>
---
 virt/kvm/kvm_main.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/virt/kvm/kvm_main.c b/virt/kvm/kvm_main.c
index 383df23514b9..3884e9f30251 100644
--- a/virt/kvm/kvm_main.c
+++ b/virt/kvm/kvm_main.c
@@ -3182,7 +3182,7 @@ static int kvm_vm_ioctl_create_vcpu(struct kvm *kvm, u32 id)
 	if (r)
 		goto vcpu_decrement;
 
-	vcpu = kmem_cache_zalloc(kvm_vcpu_cache, GFP_KERNEL);
+	vcpu = kmem_cache_zalloc(kvm_vcpu_cache, GFP_KERNEL_ACCOUNT);
 	if (!vcpu) {
 		r = -ENOMEM;
 		goto vcpu_decrement;

From patchwork Wed Mar 31 02:30:25 2021
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
X-Patchwork-Submitter: Sean Christopherson <seanjc@google.com>
X-Patchwork-Id: 12174155
Return-Path: <kvm-owner@kernel.org>
X-Spam-Checker-Version: SpamAssassin 3.4.0 (2014-02-07) on
	aws-us-west-2-korg-lkml-1.web.codeaurora.org
X-Spam-Level: 
X-Spam-Status: No, score=-26.3 required=3.0 tests=BAYES_00,DKIMWL_WL_MED,
	DKIM_SIGNED,DKIM_VALID,DKIM_VALID_AU,HEADER_FROM_DIFFERENT_DOMAINS,
	INCLUDES_CR_TRAILER,INCLUDES_PATCH,MAILING_LIST_MULTI,SPF_HELO_NONE,SPF_PASS,
	USER_AGENT_GIT,USER_IN_DEF_DKIM_WL autolearn=ham autolearn_force=no
	version=3.4.0
Received: from mail.kernel.org (mail.kernel.org [198.145.29.99])
	by smtp.lore.kernel.org (Postfix) with ESMTP id E2B17C433E1
	for <kvm@archiver.kernel.org>; Wed, 31 Mar 2021 02:31:34 +0000 (UTC)
Received: from vger.kernel.org (vger.kernel.org [23.128.96.18])
	by mail.kernel.org (Postfix) with ESMTP id B2BAE619B9
	for <kvm@archiver.kernel.org>; Wed, 31 Mar 2021 02:31:34 +0000 (UTC)
Received: (majordomo@vger.kernel.org) by vger.kernel.org via listexpand
        id S233257AbhCaCbD (ORCPT <rfc822;kvm@archiver.kernel.org>);
        Tue, 30 Mar 2021 22:31:03 -0400
Received: from lindbergh.monkeyblade.net ([23.128.96.19]:59916 "EHLO
        lindbergh.monkeyblade.net" rhost-flags-OK-OK-OK-OK) by vger.kernel.org
        with ESMTP id S233080AbhCaCae (ORCPT <rfc822;kvm@vger.kernel.org>);
        Tue, 30 Mar 2021 22:30:34 -0400
Received: from mail-qv1-xf49.google.com (mail-qv1-xf49.google.com
 [IPv6:2607:f8b0:4864:20::f49])
        by lindbergh.monkeyblade.net (Postfix) with ESMTPS id 0823AC061574
        for <kvm@vger.kernel.org>; Tue, 30 Mar 2021 19:30:34 -0700 (PDT)
Received: by mail-qv1-xf49.google.com with SMTP id i7so311666qvh.22
        for <kvm@vger.kernel.org>; Tue, 30 Mar 2021 19:30:33 -0700 (PDT)
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=google.com; s=20161025;
        h=reply-to:date:in-reply-to:message-id:mime-version:references
         :subject:from:to:cc;
        bh=9ngxz5QAuEGilGZvxilVYqHe1BwKjT3flmnusOi2/fI=;
        b=LicNR0ZkoiCXiKVmO0Ds1YFkeaxDw5kP7hrokUUJ0H/UfOMARURimbTR8ystlpw3LE
         DTG2R/6GZcZWFnTEKaTKUSI+5X/+ZROOXtoGen04R9T9IrAtnHe5lgDOqRHfhAXtu6en
         gfysf/mhPrr+0dmQPfeCQdg0G41X4orl90ynZN7tF9L5WvGZDP5jJBIHlmflCjmQBWA4
         xT+MTcqVUXUtNm7ErI84r4YaLK752It8TaLMCpDBZ/LzWUrTqBJfoXPWmBx7R5koeZoo
         SVIXqGKmsNtjyJtU7BKhIkMEEfm90yy93p5gA5Or1bwgdk+G1pjY1I7iKZ3BPML9rNSm
         A2pA==
X-Google-DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=1e100.net; s=20161025;
        h=x-gm-message-state:reply-to:date:in-reply-to:message-id
         :mime-version:references:subject:from:to:cc;
        bh=9ngxz5QAuEGilGZvxilVYqHe1BwKjT3flmnusOi2/fI=;
        b=R0WxBzFyWWK0lwowX2rU35as/JF2udmPozxt2g8xJZu72iarrc5vA57YP8zgZoN/4k
         zAfb+S1fXeH256uw2n0aU3z0QT8ZcQp0SbnDK3/dlV5iVqFIaWi3y80cSBw00KvuvYmm
         VISfds23l39SnYyA1fSSIz9bxkC+6iFagYCvjs9C+sCJ/BNz+Jtwdn8/lXNmPgWNKWeE
         /FR7KQYToYSJxy11s6NPXd34CLbVVLxb7TY5c89pSszAGmObr0auNrxnc6buOFegVyR2
         js3p1lvEd10Jp107w5V1/Y6l1H5sz+bxtfYPKIOhXbwZaKUe6fonXVBLwGlF+PF7Ffqz
         ic+Q==
X-Gm-Message-State: AOAM531gILT92C/H9mTnQkKB/BJS8StMypCfF/XfDEhfRKPLP33HUjby
        LSBInMVsAFQxTAkqqqItZf64qUkZgJA=
X-Google-Smtp-Source: 
 ABdhPJzuedRPX2SVO40eFzTyVbFitDzVxdBK989CVi+aohDw/Pi7chg4NgBOrFfsM8w4yFLDCK0+NvuKbzg=
X-Received: from seanjc798194.pdx.corp.google.com
 ([2620:15c:f:10:6c6b:5d63:9b3b:4a77])
 (user=seanjc job=sendgmr) by 2002:a0c:f092:: with SMTP id
 g18mr930054qvk.11.1617157833106;
 Tue, 30 Mar 2021 19:30:33 -0700 (PDT)
Reply-To: Sean Christopherson <seanjc@google.com>
Date: Tue, 30 Mar 2021 19:30:25 -0700
In-Reply-To: <20210331023025.2485960-1-seanjc@google.com>
Message-Id: <20210331023025.2485960-3-seanjc@google.com>
Mime-Version: 1.0
References: <20210331023025.2485960-1-seanjc@google.com>
X-Mailer: git-send-email 2.31.0.291.g576ba9dcdaf-goog
Subject: [PATCH 2/2] KVM: x86: Account a variety of miscellaneous allocations
From: Sean Christopherson <seanjc@google.com>
To: Paolo Bonzini <pbonzini@redhat.com>
Cc: Sean Christopherson <seanjc@google.com>,
        Vitaly Kuznetsov <vkuznets@redhat.com>,
        Wanpeng Li <wanpengli@tencent.com>,
        Jim Mattson <jmattson@google.com>,
        Joerg Roedel <joro@8bytes.org>, kvm@vger.kernel.org,
        linux-kernel@vger.kernel.org
Precedence: bulk
List-ID: <kvm.vger.kernel.org>
X-Mailing-List: kvm@vger.kernel.org

Switch to GFP_KERNEL_ACCOUNT for a handful of allocations that are
clearly associated with a single task/VM.

Note, there are a several SEV allocations that aren't accounted, but
those can (hopefully) be fixed by using the local stack for memory.

Signed-off-by: Sean Christopherson <seanjc@google.com>
---
 arch/x86/kvm/svm/nested.c | 4 ++--
 arch/x86/kvm/svm/sev.c    | 6 +++---
 arch/x86/kvm/vmx/vmx.c    | 2 +-
 3 files changed, 6 insertions(+), 6 deletions(-)

diff --git a/arch/x86/kvm/svm/nested.c b/arch/x86/kvm/svm/nested.c
index 8523f60adb92..4f9e8b80ef99 100644
--- a/arch/x86/kvm/svm/nested.c
+++ b/arch/x86/kvm/svm/nested.c
@@ -1234,8 +1234,8 @@ static int svm_set_nested_state(struct kvm_vcpu *vcpu,
 		return -EINVAL;
 
 	ret  = -ENOMEM;
-	ctl  = kzalloc(sizeof(*ctl),  GFP_KERNEL);
-	save = kzalloc(sizeof(*save), GFP_KERNEL);
+	ctl  = kzalloc(sizeof(*ctl),  GFP_KERNEL_ACCOUNT);
+	save = kzalloc(sizeof(*save), GFP_KERNEL_ACCOUNT);
 	if (!ctl || !save)
 		goto out_free;
 
diff --git a/arch/x86/kvm/svm/sev.c b/arch/x86/kvm/svm/sev.c
index 83e00e524513..883ce6bf23b9 100644
--- a/arch/x86/kvm/svm/sev.c
+++ b/arch/x86/kvm/svm/sev.c
@@ -637,7 +637,7 @@ static int sev_launch_measure(struct kvm *kvm, struct kvm_sev_cmd *argp)
 		}
 
 		ret = -ENOMEM;
-		blob = kmalloc(params.len, GFP_KERNEL);
+		blob = kmalloc(params.len, GFP_KERNEL_ACCOUNT);
 		if (!blob)
 			goto e_free;
 
@@ -1074,7 +1074,7 @@ static int sev_get_attestation_report(struct kvm *kvm, struct kvm_sev_cmd *argp)
 		}
 
 		ret = -ENOMEM;
-		blob = kmalloc(params.len, GFP_KERNEL);
+		blob = kmalloc(params.len, GFP_KERNEL_ACCOUNT);
 		if (!blob)
 			goto e_free;
 
@@ -1775,7 +1775,7 @@ static bool setup_vmgexit_scratch(struct vcpu_svm *svm, bool sync, u64 len)
 			       len, GHCB_SCRATCH_AREA_LIMIT);
 			return false;
 		}
-		scratch_va = kzalloc(len, GFP_KERNEL);
+		scratch_va = kzalloc(len, GFP_KERNEL_ACCOUNT);
 		if (!scratch_va)
 			return false;
 
diff --git a/arch/x86/kvm/vmx/vmx.c b/arch/x86/kvm/vmx/vmx.c
index c8a4a548e96b..5ab25376d718 100644
--- a/arch/x86/kvm/vmx/vmx.c
+++ b/arch/x86/kvm/vmx/vmx.c
@@ -562,7 +562,7 @@ static int hv_enable_direct_tlbflush(struct kvm_vcpu *vcpu)
 	 * evmcs in singe VM shares same assist page.
 	 */
 	if (!*p_hv_pa_pg)
-		*p_hv_pa_pg = kzalloc(PAGE_SIZE, GFP_KERNEL);
+		*p_hv_pa_pg = kzalloc(PAGE_SIZE, GFP_KERNEL_ACCOUNT);
 
 	if (!*p_hv_pa_pg)
 		return -ENOMEM;
